### ISSUE TITLE
TorchX Kubernetes Support

### ISSUE THREAD

**Issue by johnwhumphreys** [meta-pytorch/monarch#2039]

**Comment by punkerpunker** [comment]
That'd be awesome! Battling the torchx remote allocator + monarch now - no luck yet

**Comment by romilbhardwaj** [comment]
@johnwhumphreys I took a stab at running Monarch on k8s in #2083, curious what you think :)

**Comment by johnwhumphreys** [comment]
@romilbhardwaj  - @colin2328  will review that PR!

I just landed PRs to torchx to:
- populate hostnames
- in-cluster k8s auth
- 1 more?

**Comment by johnwhumphreys** [comment]
This now works out of box.
- TorchX mods were landed to:
    - populate k8s host names with in-cluster DNS (ip-#-#-#-#.pod.cluster.local style).
    - add in-cluster k8s auth.
    - add scheduler opts to allow enabling/disabling EFA and overriding the memory/cpu overhead (as each cluster is different based on its devops setup).
 - Monarch diffs were landed to avoid a race condition with UNKNOWN scheduler states in commands.py.
 - I pushed an alpha version container image to our public registry based on nightlies (pytorch nightly docker, torch-monarch nightly, and torchx-nightly[kubernetes] as of today and tested it.
   -  ghcr.io/meta-pytorch/monarch:2025-12-16-alpha

Tested launching that container as a Monarch client, then running a 2 node distributed DDP job from it (just in ipython).  I'll make some documentation for it next and we can try to automate a more frequent docker build as part of other work.

Here's a (very) basic YAML for running the client container.  Notice you'll have to make a service account as well that has enough permissions to execute the relevant k8s cluster scheduling work (I don't have "least privilege" one yet, so I'm going to avoid suggesting one until I do, but it needs to be able to schedule work, list work, etc).

```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: monarch-client-demo-YOURNAMEHERE
  namespace: monarch-tests
spec:
  serviceAccountName: kubectl-user
  containers:
  - name: monarch
    image: ghcr.io/meta-pytorch/monarch:2025-12-16-alpha
    # Just sleep forever
    command: [ "sleep" ]
    args: [ "infinity" ]
```
Here is an example of how to launch a 2 node job and build a proc mesh.

```python
# ------------------------------------------------------------------------------------------------------
# Step #1 - Launch a monarch job via the torchx kubernetes scheduler, which uses the volcano scheduler
#           on the cluster.  After this step, we have N pods all waiting on bootstrap with a port open.
# ------------------------------------------------------------------------------------------------------
import asyncio
import json
from getpass import getuser

from monarch.actor import enable_transport
from monarch.tools import commands
from monarch.tools.components import hyperactor

from monarch.tools.config import Config
from monarch.tools.config.workspace import Workspace

# Must be the first thing that happens after imports.
enable_transport("tcp")

# The job name needs to be prefixed with the namespace for kubernetes.
namespace = "monarch-tests"
jobname = f"{namespace}:monarch-{getuser()}"

# General job config.
volcano_queue = "monarch"
image = "ghcr.io/meta-pytorch/monarch:2025-12-16-alpha"
mesh_name = "mesh0"
num_hosts = 2
num_gpus_per_host = 4

# Must be one of https://github.com/meta-pytorch/torchx/blob/main/torchx/specs/named_resources_aws.py#L421.
# Use overrides in scheduler_args to tune the overhead, disable EFA if needed.
host_type = "aws_g5.12xlarge"



async def schedule_monarch():
    config = Config(
        scheduler="k8s",
        scheduler_args={
            "queue": volcano_queue,
            "image_repo": image,
            "namespace": namespace,
            "reserved_millicpu": 3000,
            "efa_device_count": 0,
        },
        appdef=hyperactor.host_mesh(
            image=image, meshes=[f"{mesh_name}:{num_hosts}:{host_type}"], env={}
        ),
        workspace=Workspace(env=None),
    )

    # set to True to force kill the job and restart it even if it exists
    restart = False

    # Try to get status of job, if it returns none, it doesn't exist, then it submits a job.
    # If it comes back with a non-null status; wait to see the job running, then return.
    # Server info has all info about how to connect to individual process allocators on the job.
    server_info = await commands.get_or_create(jobname, config, force_restart=restart)
    print(f"""
   ===== Server Info ====="
   {json.dumps(server_info.to_json(), indent=2)}
   """)
    return server_info


# Run the async function in its own loop so we can await commands.get_or_create()...
server_info = asyncio.run(schedule_monarch())

# ------------------------------------------------------------------------------------------------------
# Step #2 - Use a static remote allocator to allocate a proc mesh on the servers prepared above.
# ------------------------------------------------------------------------------------------------------
from monarch._rust_bindings.monarch_hyperactor.alloc import AllocConstraints, AllocSpec
from monarch._src.actor.allocator import RemoteAllocator, StaticRemoteAllocInitializer
from monarch.actor import ProcMesh

# get the host addrs of all the hosts in the mesh
mesh_info = server_info.get_mesh_spec(mesh_name)
host_addrs = mesh_info.server_addrs()

# static alloc initializer just takes a list of host addrs to comprise the host mesh
alloc_initializer = StaticRemoteAllocInitializer(*host_addrs)
world_id = f"{server_info.name}_{mesh_name}"
allocator = RemoteAllocator(world_id=world_id, initializer=alloc_initializer)
mesh_dimensions = {"hosts": num_hosts, "gpus": num_gpus_per_host}
spec = AllocSpec(AllocConstraints(), **mesh_dimensions)
alloc = allocator.allocate(spec)
proc_mesh = ProcMesh.from_alloc(alloc)
```
Here is an example job to run (DDP, multinode).

```python
import os
import socket

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim

from monarch.actor import Actor, current_rank, endpoint
from monarch.utils import setup_env_for_distributed

from torch.nn.parallel import DistributedDataParallel as DDP

WORLD_SIZE = 8
GPUS_PER_HOST = 4


class ToyModel(nn.Module):
    """A simple toy model for demonstration purposes."""

    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


class DDPActor(Actor):
    def __init__(self):
        """Initialize the DDP actor.

        Note: We don't take master_addr/master_port as parameters anymore.
        Instead, we rely on setup_env_for_distributed() to set the environment variables.
        """
        rank = current_rank()
        self.rank = rank.rank

        # Get the local rank from the environment variable set by setup_env_for_distributed
        # This is the proper way to get local_rank in Monarch
        self.local_rank = int(os.environ.get("LOCAL_RANK", rank["gpus"]))

        # CRITICAL: Set CUDA device IMMEDIATELY in __init__, before any CUDA operations
        # This must happen before ANY torch.cuda calls that might initialize CUDA
        # If CUDA is initialized before set_device, NCCL may detect the wrong GPU
        import torch

        if torch.cuda.is_available():
            torch.cuda.set_device(self.local_rank)

        # Debug: print rank information
        hostname = socket.gethostname()
        print(
            f"[INIT] rank={self.rank}, local_rank={self.local_rank}, "
            f"hostname={hostname}, "
            f"CUDA_VISIBLE_DEVICES={os.environ.get('CUDA_VISIBLE_DEVICES', 'NOT SET')}, "
            f"MASTER_ADDR={os.environ.get('MASTER_ADDR', 'NOT SET')}, "
            f"current_device={torch.cuda.current_device() if torch.cuda.is_available() else 'N/A'}"
        )

    def _rprint(self, msg):
        """Helper method to print with rank information."""
        hostname = socket.gethostname()
        print(f"[rank={self.rank} local_rank={self.local_rank} host={hostname}] {msg}")

    @endpoint
    async def setup(self):
        """Initialize the PyTorch distributed process group."""
        # Read environment variables set by setup_env_for_distributed
        master_addr = os.environ.get("MASTER_ADDR")
        master_port = os.environ.get("MASTER_PORT")
        world_size = int(os.environ.get("WORLD_SIZE", WORLD_SIZE))

        self._rprint(
            f"Environment: MASTER_ADDR={master_addr}, MASTER_PORT={master_port}"
        )

        # CRITICAL: Set the CUDA device BEFORE any CUDA queries
        # Even torch.cuda.device_count() can initialize CUDA!
        self._rprint(f"Setting CUDA device to local_rank={self.local_rank}")
        torch.cuda.set_device(self.local_rank)

        # Verify it was set correctly
        actual_device = torch.cuda.current_device()
        self._rprint(f"Verified CUDA device is now: {actual_device}")

        # Now it's safe to query device count
        self._rprint(f"Number of available GPUs: {torch.cuda.device_count()}")

        # NCCL configuration for better stability
        os.environ["NCCL_DEBUG"] = "INFO"
        
        # Disable InfiniBand if not available (common in cloud environments)
        os.environ["NCCL_IB_DISABLE"] = "1"

        # CRITICAL: Set socket family to force IPv4
        # This can help with hostname resolution issues in K8s
        os.environ["NCCL_SOCKET_FAMILY"] = "AF_INET"

        # CRITICAL: Set NCCL_HOSTID to differentiate between hosts
        # This prevents NCCL from thinking processes on different hosts are on the same host
        # Without this, NCCL may compute the same hostHash for different physical hosts
        group_rank = int(os.environ.get("GROUP_RANK", self.rank // GPUS_PER_HOST))
        hostname = socket.gethostname()
        nccl_hostid = f"{hostname}_host_{group_rank}"
        os.environ["NCCL_HOSTID"] = nccl_hostid
        self._rprint(f"Set NCCL_HOSTID={nccl_hostid} (group_rank={group_rank})")

        # Print all NCCL-related environment variables for debugging
        import subprocess

        try:
            hostname_output = subprocess.check_output(["hostname"]).decode().strip()
            hostname_i_output = (
                subprocess.check_output(["hostname", "-I"]).decode().strip()
            )
            self._rprint(f"Hostname: {hostname_output}")
            self._rprint(f"Hostname -I: {hostname_i_output}")
        except Exception as e:
            self._rprint(f"Could not get hostname info: {e}")

        self._rprint(
            f"Calling init_process_group (rank={self.rank}, world_size={world_size})"
        )

        # Initialize the process group with timeout
        try:
            dist.init_process_group(
                "nccl",
                rank=self.rank,
                world_size=world_size,
                timeout=torch.distributed.timedelta(seconds=180),
            )
            self._rprint("✓ Successfully initialized torch distributed")
        except Exception as e:
            self._rprint(f"✗ Failed to initialize: {e}")
            raise

    @endpoint
    async def cleanup(self):
        """Clean up the PyTorch distributed process group."""
        self._rprint("Cleaning up torch distributed")
        dist.destroy_process_group()

    @endpoint
    async def demo_basic(self):
        """Run a basic DDP training example."""
        self._rprint("Running basic DDP example")

        # CRITICAL: Re-set the CUDA device to ensure it's correct for this process
        torch.cuda.set_device(self.local_rank)

        # Verify the device is set correctly
        current_device = torch.cuda.current_device()
        self._rprint(
            f"Current CUDA device: {current_device}, expected: {self.local_rank}"
        )

        # Create model and move it to GPU with local_rank
        device = torch.device(f"cuda:{self.local_rank}")
        model = ToyModel().to(device)
        ddp_model = DDP(model, device_ids=[self.local_rank])

        loss_fn = nn.MSELoss()
        optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)

        optimizer.zero_grad()
        outputs = ddp_model(torch.randn(20, 10).to(device))
        labels = torch.randn(20, 5).to(device)
        loss_fn(outputs, labels).backward()
        optimizer.step()

        self._rprint("✓ Finished running basic DDP example")

    @endpoint
    async def test_connectivity(self):
        """Test network connectivity to master node."""
        hostname = socket.gethostname()
        master_addr = os.environ.get("MASTER_ADDR", "NOT SET")
        master_port = os.environ.get("MASTER_PORT", "NOT SET")

        self._rprint(f"Testing connectivity from {hostname}")

        # Try to resolve the master address
        try:
            master_ip = socket.gethostbyname(master_addr)
            self._rprint(f"Master address '{master_addr}' resolves to {master_ip}")
        except Exception as e:
            self._rprint(f"✗ Cannot resolve master address '{master_addr}': {e}")

        # Try to connect to the master port
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(5)
            result = sock.connect_ex((master_addr, int(master_port)))
            sock.close()
            if result == 0:
                self._rprint(f"✓ Can connect to {master_addr}:{master_port}")
            else:
                self._rprint(
                    f"✗ Cannot connect to {master_addr}:{master_port} (error {result})"
                )
        except Exception as e:
            self._rprint(f"✗ Connection test failed: {e}")


async def create_ddp_actors(proc_mesh):
    """Create the process mesh and spawn DDP actors.

    Note: No need to pass master_addr/master_port - setup_env_for_distributed handles this.
    """
    # Spawn our actor mesh on top of the process mesh
    ddp_actor = proc_mesh.spawn("ddp_actor", DDPActor)
    return ddp_actor


async def test_connectivity(ddp_actor):
    """Test network connectivity before initializing distributed."""
    print("\n=== Testing network connectivity ===")
    await ddp_actor.test_connectivity.call()
    print("=== Connectivity test complete ===\n")


async def setup_distributed(ddp_actor):
    """Initialize the distributed environment."""
    print("\n=== Setting up distributed training ===")
    await ddp_actor.setup.call()
    print("=== Distributed setup complete ===\n")


async def run_ddp_example(ddp_actor):
    """Run the DDP training example."""
    print("\n=== Running DDP example ===")
    await ddp_actor.demo_basic.call()
    print("=== DDP example complete ===\n")


async def cleanup_distributed(ddp_actor):
    """Clean up distributed resources."""
    print("\n=== Cleaning up ===")
    await ddp_actor.cleanup.call()
    print("=== Cleanup complete ===\n")


async def main(proc_mesh, mesh_info):
    """Main function to run the DDP example.

    Args:
        proc_mesh: The process mesh spanning 2 hosts with 4 GPUs each
        mesh_info: Mesh info containing hostnames and other metadata
    """
    # Print mesh info for debugging
    print(f"Mesh hostnames: {mesh_info.hostnames}")
    print(f"Using master addr (rank 0 host): {mesh_info.hostnames[0]}")

    master_addr = mesh_info.hostnames[0]

    # CRITICAL: Setup environment variables for distributed training
    # This must be done BEFORE spawning the DDP actors!
    print("\n=== Setting up environment for distributed training ===")
    await setup_env_for_distributed(
        proc_mesh, master_addr=master_addr, master_port=29500
    )
    print("=== Environment setup complete ===\n")

    # Create actors (after setup_env_for_distributed)
    ddp_actor = await create_ddp_actors(proc_mesh)

    # Test connectivity first (helps debug networking issues)
    await test_connectivity(ddp_actor)

    # Setup distributed environment
    await setup_distributed(ddp_actor)

    # Run DDP example
    await run_ddp_example(ddp_actor)

    # Clean up
    await cleanup_distributed(ddp_actor)

    print("DDP example completed successfully!")


if __name__ == "__main__":
    import asyncio

    # Example: asyncio.run(main(proc_mesh, mesh_info))
    asyncio.run(main(proc_mesh, mesh_info))
```

We likely should make a JobTrait for this integration next.
